clang/test/SemaSYCL/intel-max-work-group-size.cpp

```diff
@@ -101,10 +101,10 @@ f15() {} // OK
 [[intel::max_work_group_size(1, 2, 3)]] [[sycl::reqd_work_group_size(1, 2, 3)]] void f17(){}; // OK
 
 [[sycl::reqd_work_group_size(16)]]            // expected-note {{conflicting attribute is here}}
-[[intel::max_work_group_size(1, 1, 16)]] void // expected-error {{'max_work_group_size' attribute conflicts with 'reqd_work_group_size' attribute}}
+[[intel::max_work_group_size(16, 1, 1)]] void // expected-error {{'max_work_group_size' attribute conflicts with 'reqd_work_group_size' attribute}}
 f18();
 
-[[intel::max_work_group_size(16, 16, 1)]] void f19();
+[[intel::max_work_group_size(1, 16, 16)]] void f19();
 [[sycl::reqd_work_group_size(16, 16)]] void f19(); // OK
```

Comment:
This is because `max_work_group_size`, like `reqd_work_group_size`, is flipped, but since it always has 3 arguments it is X and Z it swaps, while the `reqd_work_group_size` here only has two arguments, so it only swaps those.

> What is backend workgroup size corresponding to sycl::reqd_work_group_size(3,4) and how does it relate to checks with max_work_group_size?

`sycl::reqd_work_group_size(3,4)` corresponds to `<4,3,1>` which would correspond to `max_work_group_size(1,3,4)`. Honestly I would prefer changing `max_work_group_size` too to avoid this kind of confusion, but it is a little more complicated because I am not sure we want to restrict the dimensionality of it necessarily, like we do with `reqd_work_group_size`. Index flipping is the gift that keeps on giving.

Reply:
No matter how many times I try to make sense of this, I don't think I ever will.  :-)

Reply:
>>In my opinion we should change the intel::max_work_group_size attribute to allow 1-3 arguments too. That would make it easier for users to actually use this with the other work-group-related attributes and the actual work-group sizes specified for kernels at launch.

+1 from me

Reply:
No objections from me.  If there are concerns about breaking existing code, my suggestion would be to use a warning for the attribute spelling and an error for the eventual compile-time properties spelling.

Reply:
>>In my opinion we should change the intel::max_work_group_size attribute to allow 1-3 arguments too. That would make it easier for users to actually use this with the other work-group-related attributes and the actual work-group sizes specified for kernels at launch.

@steffenlarsen, could you please create an issue about it so that we do not forget this discussion? You can assign to me if you do not plan to work on it.

Reply:
https://github.com/intel/llvm/issues/7700